### PR TITLE
Use commands for container push

### PR DESCRIPTION
**Pull request by bangelic**
Containers are able to be pushed to Satellite's
container registry.
This command for pushing containers was needed to show the user how to
do so.

* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [ ] Foreman 3.10/Katello 4.12
* [ ] Foreman 3.9/Katello 4.11 (Satellite 6.15; orcharhino 6.8)
* [ ] Foreman 3.8/Katello 4.10
* [ ] Foreman 3.7/Katello 4.9 (Satellite 6.14)
* [ ] Foreman 3.6/Katello 4.8
* [ ] Foreman 3.5/Katello 4.7 (Satellite 6.13; orcharhino 6.6/6.7)
* [ ] Foreman 3.4/Katello 4.6 (EL8 only)
* [ ] Foreman 3.3/Katello 4.5 on EL7 & EL8 (Satellite 6.12 on EL8 only; orcharhino 6.4/6.5 on EL8 only)
* We do not accept PRs for Foreman older than 3.3.

### CODE REVIEW

**Review by asteflova — COMMENTED** [review]
First pass at a review :) I'm making a few suggestions that would help integrate the new additions more organically with the existing content of the module you're updating. Let me know if I'm misinterpreting something.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (bangelic) [reply guides/common/modules/proc_using-container-registries.adoc:36]: @asteflova I think the introduction of the module makes sense since, as you previously mentioned, the module's introduction covers only the previous scope of content.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:6]:
```suggestion
The {Project} registry follows the Open Containers Initiative (OCI) specification, so you can push content to Katello by using the same methods that apply to other registries.
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:48]:
```suggestion
* An example of a complete push:
```
The meaning is basically the same, but using the keyword 'example' would really help users scan the text more quickly.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:44]: You should use _italics_ for example values that are to be replaced by users. I supposed that would look like this:
```suggestion
# podman push _image_ _example.com_/_organization_label_/_product_label_/_repository_name_[:tag]
```

Comment (asteflova) [reply guides/common/modules/proc_using-container-registries.adoc:44]: Actually, the existing example commands in this module use {foreman-example-com}. Which would translate to:

```suggestion
# podman push _image_ _{foreman-example-com}_/_organization_label_/_product_label_/_repository_name_[:tag]
```

Comment (asteflova) [reply guides/common/modules/proc_using-container-registries.adoc:44]: Please apply the italics and the attribute for example.com to your other command examples too.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:61]:
```suggestion
Subsequent pushes to the same repository update the contents of that repository.
```
This line and the ones above are tightly related so one bullet point would be enough.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:65]:
```suggestion
* The pushed container name must contain only lowercase characters.
```
I suppose you don't really need to mention Podman. The sentence is more straightforward without it.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:68]:
```suggestion
* Unless pushed repositories are published in a content view version, they do not follow the Registry Name Pattern.
This is to ensure that users can push and pull from the same path.
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:64]:
```suggestion
* You can only push content to the {ProjectServer} itself.
If you need pushed content on {SmartProxies} as well, use {SmartProxy} syncing.
```

Comment (asteflova) [reply guides/common/modules/proc_using-container-registries.adoc:64]: Line 64 immediately raised a question for me: How do I use SmartProxy syncing? Does the sync happen automatically, or do I start it manually?

If it happens automatically, I think you could actually drop this line. If it needs to be started manually, it would be good to explain how or add a link to another part of the docs that explains it.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:71]:
```suggestion
* Users are required to push and pull from the same path.
If a user uses the label-based schema, they pull using labels.
If a user uses the ID-based schema, they pull using IDs.
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [reply guides/common/modules/proc_using-container-registries.adoc:61]: But then again, do you really _need_ this line? I don't know much about container repositories, but doesn't pushing to the repository always mean that the repository gets updated? :shrug: I suggest to drop the line.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:60]:
```suggestion
```
I believe you don't need this line.

If Katello sends an error message about that and that message is helpful enough, users will know. (If the message is _not_ helpful enough, then the right thing to do is to fix that message rather than explain it's meaning in the docs.)

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:66]: I've seen information about user permissions included under a `.Prerequisite` heading in other modules. Can you create a heading like that and move this line there?

Also, how do I know that my user is able to create repositories? Are we talking about Foreman roles/permissions? Which ones?

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:59]:
```suggestion
* A repository in Katello is created after the content is pushed.
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:62]: This is curious. The procedure mentions pulling container images manually. Does this mean that when you push a container image, that image gets automatically pulled afterwards?

I don't know what to do about this sentence, but I felt the need to mention that it confused me in the context of the whole module :shrug:

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:58]: Can you move this subheading for it to be placed somewhere closer to the beginning of the module? I don't think the information here will be relevant after `.Procedure`, after users have run the command(s).

Comment (asteflova) [reply guides/common/modules/proc_using-container-registries.adoc:58]: Also, feel free to reevaluate whether including the limitations as bullet points still works if/after you implement my suggestions. It's quite possible that after the edits, regular paragraphs will make more sense.

**Review by asteflova — COMMENTED** [review]
Thanks for working on this @bangelic, the changes you've made so far made it a lot easier for me to understand the context of your addition more clearly.

In this round of feedback, I'm offering mostly specific suggestions, but I also have a couple of open questions in places where I had trouble understanding what was being described.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (bangelic) [reply guides/common/modules/proc_using-container-registries.adoc:64]: @ianballou Does SmartProxy syncing happen automatically or does the user need to do it manually?

File: guides/common/modules/proc_using-container-registries.adoc
Comment (bangelic) [reply guides/common/modules/proc_using-container-registries.adoc:66]: @ianballou Does a user know if they have permissions to create repositories? Specific role or permissions needed? I am pushing a commit to say that a user must have permissions to create repositories in order to push content to an organization or product.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (bangelic) [reply guides/common/modules/proc_using-container-registries.adoc:62]: @ianballou Any chance you can provide some clarity here?

File: guides/common/modules/proc_using-container-registries.adoc
Comment (ianballou) [reply guides/common/modules/proc_using-container-registries.adoc]: A user indeed must have the same permissions necessary to create a repository. For creation, that would be `edit_products`.

Pulling content is the same way as long as the lifecycle environment in question has "unauthenticated pull" turned off (which is default).

Pull permissions: `view_lifecycle_environments`, `view_products`, `view_content_views`.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (ianballou) [inline guides/common/modules/proc_using-container-registries.adoc:20]: This is a statement on how the feature works rather than a limitation. Perhaps it should go down in the procedure to tell users to expect that a repository gets created in Katello.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (ianballou) [inline guides/common/modules/proc_using-container-registries.adoc:19]: Maybe something like:
```suggestion
.Limitations of pushing content to the {Project} container registry
```
The limitations that we're noting here are outside of the usual registry specification.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (ianballou) [inline guides/common/modules/proc_using-container-registries.adoc:52]: "Container namespaces" isn't in the OCI container spec, so I think we can leave that term out.
```suggestion
* To indicate which organization, product, and repository the container image belongs to, include the organization and product in the container repository name.
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (ianballou) [inline guides/common/modules/proc_using-container-registries.adoc:21]: This was supposed to be:
```suggestion
* Content is pullable as soon as it is pushed.
```
which was just a way of saying that there shouldn't be any delay between pushing content and being able to pull it. I think it might be extra info that's not needed in the docs.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (ianballou) [reply guides/common/modules/proc_using-container-registries.adoc]: I added my thoughts on this to a review, let me know if there are lingering questions.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:22]:
```suggestion
.Considerations for pushing content to the {Project} container registry
```
or

```suggestion
.Guidelines for pushing content to the {Project} container registry
```

I'm thinking of how to frame this in a more neutral light, to avoid making it sound like the feature is not working properly. What do you think @bangelic, would this describe the purpose of the list accurately enough?

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:74]:
```suggestion
A repository in Katello is created after the content is pushed.
```
I think this would work better as a separate para, to conclude the section on pushing.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:57]: Is this about an action that users must take? Like "Compose all labels entirely of lowercase chars"?

I'm asking because if this an action item for users, it should be written in imperative voice. If it's just an FYI item, then perhaps it doesn't need to be included at all..? It's a bit confusing: I'm reading this and I don't know if I'm being asked to do something here.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:59]:
```suggestion
+
[options="nowrap", subs="+quotes,attributes"]
```
The contents of the command blocks don't render properly. I think it might be because a plus sign is missing, can you please check? The same would also apply to the block below.

**Review by asteflova — CHANGES_REQUESTED** [review]

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:5]:
```suggestion
Containers are pushed to the {Project} container registry.
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:7]:
```suggestion
For more information about OCI, see link:https://opencontainers.org/[Open Container Initiative Distribution Specification].
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (bangelic) [reply guides/common/modules/proc_using-container-registries.adoc:22]: @asteflova I think I like **Considerations** better :)

File: guides/common/modules/proc_using-container-registries.adoc
Comment (bangelic) [reply guides/common/modules/proc_using-container-registries.adoc:57]: It isn't an action by the user so much as that is how the labels are composed. I will remove it.

**Comment by asteflova** [comment]
Hi @bangelic I'd like to see one more tweak before I approve this PR! :) If you look at how the procedure is currently structured, you'll see that the first three items are numbered but the last one (pushing -- your addition) is a regular para.

Screenshot from 2024-06-26 07-28-56

That looks extremely weird and could confuse users. Similar items (logging, listing, pulling, pushing) should be structured in the same way: for example they should all be ordered list items or they should all be regular paragraphs. Can you please take a look at this?

**Comment by bangelic** [comment]
@asteflova I made the last step a number and the remainder of that step has bullet points. Let me know what you think.

File: guides/common/modules/proc_using-container-registries.adoc
Comment (asteflova) [inline guides/common/modules/proc_using-container-registries.adoc:5]:
```suggestion
Containers are pushed to the {Project} container registry.
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (Lennonka) [inline guides/common/modules/proc_using-container-registries.adoc:72]:
```suggestion
```
We don't include outputs. I'm not sure why. And if we remove it, does it still make sense to include the example? Do we really need it?

File: guides/common/modules/proc_using-container-registries.adoc
Comment (Lennonka) [inline guides/common/modules/proc_using-container-registries.adoc:60]:
```suggestion
# podman push _image_ _{foreman-example-com}_/_organization_label_/_product_label_/_repository_name_[:_tag_]
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (Lennonka) [inline guides/common/modules/proc_using-container-registries.adoc:56]:
```suggestion
* Push accepts the following formats:
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (Lennonka) [inline guides/common/modules/proc_using-container-registries.adoc:5]:
```suggestion
You can use Podman and Docker to fetch content from container registries and push the content to the {Project} container registry.
```

File: guides/common/modules/proc_using-container-registries.adoc
Comment (Lennonka) [inline guides/common/modules/proc_using-container-registries.adoc:6]:
```suggestion
The {Project} registry follows the Open Containers Initiative (OCI) specification, so you can push content to {Project} by using the same methods that apply to other registries.
```

**Review by Lennonka — CHANGES_REQUESTED** [review]
A couple of suggestions. It would be good to remake it into a proper procedure.

**Comment by asteflova** [comment]
> A couple of suggestions. It would be good to remake it into a proper procedure.

For the record, in my review, I focused on adding the new commands and I relied on the existing supporting structure. The current structure is not ideal (it's basically a reference module, not a procedure), but I think that considering the scope of this PR, it's enough. If there is a need to rework the entire module (into a proper procedure, into a reference module, ...), that should IMO be addressed in another PR.

**Comment by Lennonka** [comment]
@bangelic You might have forgotten a pull at some point after committing changes in the GitHub web UI or something. But it's no biggie to open a new PR for these changes and continue there. You can go ahead.

**Pull request closed without merging**
